File: src/main/java/hulio13/articlesApi/domain/service/AuthorService.java
package hulio13.articlesApi.domain.service;

import hulio13.articlesApi.domain.entity.Article;
import hulio13.articlesApi.domain.entity.Author;
import hulio13.articlesApi.domain.data.repository.AuthorRepository;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
@Transactional
public class AuthorService {
    private final AuthorRepository repository;

    public AuthorService(AuthorRepository repository) {
        this.repository = repository;
    }

    public void add(Author author){
        repository.create(author);
    }

    public void update(Author author){
        repository.update(author);
    }

    public Optional<Author> getById(long id){
        return repository.getById(id);
    }

    public Optional<Author> getByAuthorName(String authorName){
        return repository.getByAuthorName(authorName);
    }

    public void remove(Author entity){
        repository.remove(entity);
    }

    public void removeById(long id) {
        repository.removeById(id);
    }

    public void linkArticleToAuthor(Author author, Article article) {
        author.addArticle(article);
        article.addAuthor(author);
        repository.update(author);
    }
}

File: src/main/java/hulio13/articlesApi/domain/service/AuthorService.kt
package hulio13.articlesApi.domain.service

import hulio13.articlesApi.application.dto.AuthorDto
import hulio13.articlesApi.domain.data.repository.ArticleRepository
import hulio13.articlesApi.domain.data.repository.AuthorRepository
import hulio13.articlesApi.domain.entity.Article
import hulio13.articlesApi.domain.entity.Author
import hulio13.articlesApi.domain.entity.author.AuthorName
import hulio13.articlesApi.domain.entity.common.Result
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.util.*

@Service
@Transactional
class AuthorService(
    private val authorRepository: AuthorRepository,
    private val articleRepository: ArticleRepository
    ) {
    fun add(dto: AuthorDto): Result<Author> {
        if (dto.name == null) return Result.error(1, "Invalid data, name must not be empty.")

        val author = Author(null, AuthorName(dto.name!!))

        return Result.ok(authorRepository.create(author))
    }

    fun update(author: Author) {
        authorRepository.update(author)
    }

    fun getById(id: Long): Optional<Author> {
        return authorRepository.getById(id)
    }

    fun getByAuthorName(authorName: String?): Optional<Author> {
        return authorRepository.getByAuthorName(authorName)
    }

    fun remove(entity: Author) {
        authorRepository.remove(entity)
    }

    fun removeById(id: Long) {
        authorRepository.removeById(id)
    }

    fun linkArticleToAuthor(author: Author, article: Article) {
        author.articles.add(article)
        article.authors.add(author)
        authorRepository.update(author)
    }
}
